bot.js

```diff
@@ -172,19 +178,23 @@ function formatDifferenceMessage(difference){
 
 function balance(pool, message){
 
-		/**
-		TODO:
-		Reintroduce Bishop logic
-		Allow leaders to be promoted/demoted and rebalance groups
-		**/
-
-		let [joining, weakest, total] = [pool.slice().sort((a,b) => a.rank-b.rank), [{rank: 100}], [...leaders, ...pool].length]
+		let [bishops, joining, weakest, total] = [pool.filter(member => member.role === "bishop").sort((a,b) => a.rank-b.rank), pool.slice().sort((a,b) => a.rank-b.rank), [{rank: 100}], [...leaders, ...pool].length]
 		leaders.forEach((leader, index) => total > (10*index) ? groups[index] = [leader] : joining.push(leader))
 		while(joining.length){
-			groups.forEach(group => {
+			groups.filter(group => !group.full).forEach(group => {
```

Comment:
Dunno how full works but looks good.

Reply:
Undefined and false are synonymous in JS - defining the value and setting it to true if a group reaches 10 and continuing in the loop